In [1]:
# 1. Installing the  Required Tools
!apt-get install -y sra-toolkit fastqc trimmomatic
!pip install cutadapt


Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  blends-common default-jre default-jre-headless fonts-dejavu-core
  fonts-dejavu-extra libapache-pom-java libargs4j-java libatk-wrapper-java
  libatk-wrapper-java-jni libcommons-compress-java libcommons-io-java
  libcommons-jexl2-java libcommons-lang3-java libcommons-logging-java
  libcommons-math3-java libcommons-parent-java libfindbin-libs-perl
  libhtsjdk-java libjbzip2-java libjson-simple-java libkdf5-2 libncbi-vdb2
  libncbi-wvdb2 libngs-java libngs-sdk-dev libngs-sdk2 libsis-base-java
  libsis-base-jni libsis-jhdf5-java libsis-jhdf5-jni libsnappy-java
  libsnappy-jni libxtst6 libxxf86dga1 med-config menu openjdk-11-jre x11-utils
Suggested packages:
  blends-doc libxz-java libcommons-io-java-doc libcommons-lang3-java-doc
  libavalon-framework-java libcommons-logging-java-doc
  libexcalibur-logkit-java liblog4j1.2-java picard-tools l

In [2]:
# 2. Verifying the  Tool Versions
!fastqc -v
!trimmomatic -version
!cutadapt --version
!fastq-dump --version


FastQC v0.11.9
/bin/bash: line 1: trimmomatic: command not found
5.2

"fastq-dump" version 2.11.3



In [3]:
# 3. Downloading the  SRA Data from NCBI Database
!mkdir -p /content/sra_data
%cd /content/sra_data

# Download and split paired-end reads
!fastq-dump --split-files SRR531365 --gzip

# Confirm files
!ls -lh /content/sra_data


/content/sra_data
Read 3154678 spots for SRR531365
Written 3154678 spots for SRR531365
total 379M
-rw-r--r-- 1 root root 191M Oct 30 04:26 SRR531365_1.fastq.gz
-rw-r--r-- 1 root root 189M Oct 30 04:26 SRR531365_2.fastq.gz


In [4]:
# 4. Running  FastQC
!mkdir -p /content/fastqc_pre
!fastqc SRR531365_1.fastq.gz SRR531365_2.fastq.gz -o /content/fastqc_pre
!ls -lh /content/fastqc_pre

from google.colab import files
files.download("/content/fastqc_pre/SRR531365_1_fastqc.html")
files.download("/content/fastqc_pre/SRR531365_2_fastqc.html")


Started analysis of SRR531365_1.fastq.gz
Approx 5% complete for SRR531365_1.fastq.gz
Approx 10% complete for SRR531365_1.fastq.gz
Approx 15% complete for SRR531365_1.fastq.gz
Approx 20% complete for SRR531365_1.fastq.gz
Approx 25% complete for SRR531365_1.fastq.gz
Approx 30% complete for SRR531365_1.fastq.gz
Approx 35% complete for SRR531365_1.fastq.gz
Approx 40% complete for SRR531365_1.fastq.gz
Approx 45% complete for SRR531365_1.fastq.gz
Approx 50% complete for SRR531365_1.fastq.gz
Approx 55% complete for SRR531365_1.fastq.gz
Approx 60% complete for SRR531365_1.fastq.gz
Approx 65% complete for SRR531365_1.fastq.gz
Approx 70% complete for SRR531365_1.fastq.gz
Approx 75% complete for SRR531365_1.fastq.gz
Approx 80% complete for SRR531365_1.fastq.gz
Approx 85% complete for SRR531365_1.fastq.gz
Approx 90% complete for SRR531365_1.fastq.gz
Approx 95% complete for SRR531365_1.fastq.gz
Analysis complete for SRR531365_1.fastq.gz
Started analysis of SRR531365_2.fastq.gz
Approx 5% complete fo

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [5]:
#  5. Preparing the  Trimmomatic Adapter File
!rm -rf /usr/local/bin/trimmomatic*
!rm -rf trimmomatic-0.39.zip Trimmomatic-0.39
!wget -q https://github.com/usadellab/Trimmomatic/releases/download/v0.39/trimmomatic-0.39.zip
!unzip -q trimmomatic-0.39.zip
!mv Trimmomatic-0.39 /usr/local/bin/
!echo 'java -jar /usr/local/bin/Trimmomatic-0.39/trimmomatic-0.39.jar "$@"' > /usr/local/bin/trimmomatic
!chmod +x /usr/local/bin/trimmomatic

!mkdir -p /usr/local/bin/Trimmomatic-0.39/adapters
!wget -q https://raw.githubusercontent.com/timflutre/trimmomatic/master/adapters/TruSeq3-PE.fa -O /usr/local/bin/Trimmomatic-0.39/adapters/TruSeq3-PE.fa


In [6]:
# 6. Trimming the Reads with Trimmomatic tool
!mkdir -p /content/trimmed
!trimmomatic PE -threads 2 \
SRR531365_1.fastq.gz SRR531365_2.fastq.gz \
/content/trimmed/SRR531365_1_paired.fq.gz /content/trimmed/SRR531365_1_unpaired.fq.gz \
/content/trimmed/SRR531365_2_paired.fq.gz /content/trimmed/SRR531365_2_unpaired.fq.gz \
ILLUMINACLIP:/usr/local/bin/Trimmomatic-0.39/adapters/TruSeq3-PE.fa:2:30:10 \
LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36


TrimmomaticPE: Started with arguments:
 -threads 2 SRR531365_1.fastq.gz SRR531365_2.fastq.gz /content/trimmed/SRR531365_1_paired.fq.gz /content/trimmed/SRR531365_1_unpaired.fq.gz /content/trimmed/SRR531365_2_paired.fq.gz /content/trimmed/SRR531365_2_unpaired.fq.gz ILLUMINACLIP:/usr/local/bin/Trimmomatic-0.39/adapters/TruSeq3-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36
Using PrefixPair: 'TACACTCTTTCCCTACACGACGCTCTTCCGATCT' and 'GTGACTGGAGTTCAGACGTGTGCTCTTCCGATCT'
ILLUMINACLIP: Using 1 prefix pairs, 0 forward/reverse sequences, 0 forward only sequences, 0 reverse only sequences
Quality encoding detected as phred33
Input Read Pairs: 3154678 Both Surviving: 482081 (15.28%) Forward Only Surviving: 137077 (4.35%) Reverse Only Surviving: 15159 (0.48%) Dropped: 2520361 (79.89%)
TrimmomaticPE: Completed successfully


In [7]:
# 7. Trimming the  Reads with Cutadapt tool
!mkdir -p /content/cutadapt_out
!cutadapt -a AGATCGGAAGAGC -A AGATCGGAAGAGC \
-o /content/cutadapt_out/SRR531365_1_trimmed.fq.gz \
-p /content/cutadapt_out/SRR531365_2_trimmed.fq.gz \
SRR531365_1.fastq.gz SRR531365_2.fastq.gz


This is cutadapt 5.2 with Python 3.12.12
Command line parameters: -a AGATCGGAAGAGC -A AGATCGGAAGAGC -o /content/cutadapt_out/SRR531365_1_trimmed.fq.gz -p /content/cutadapt_out/SRR531365_2_trimmed.fq.gz SRR531365_1.fastq.gz SRR531365_2.fastq.gz
Processing paired-end reads on 1 core ...
Done           00:00:30     3,154,678 reads @   9.6 µs/read;   6.26 M reads/minute

=== Summary ===

Total read pairs processed:          3,154,678
  Read 1 with adapter:                 125,936 (4.0%)
  Read 2 with adapter:                 101,301 (3.2%)
Pairs written (passing filters):     3,154,678 (100.0%)

Total basepairs processed:   637,244,956 bp
  Read 1:   318,622,478 bp
  Read 2:   318,622,478 bp
Total written (filtered):    631,397,995 bp (99.1%)
  Read 1:   314,515,227 bp
  Read 2:   316,882,768 bp

=== First read: Adapter 1 ===

Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 125936 times

Minimum overlap: 3
No. of allowed errors:
1-9 bp: 0; 10-13 bp: 1

Bases preceding remov

Here is an interpretation of the Cutadapt output from the previous cell:

*   **`Total read pairs processed: 3,154,678`**: This is the total number of paired-end reads that were analyzed by Cutadapt.
*   **`Read 1 with adapter: 125,936 (4.0%)`**: This indicates that 125,936 reads in the first file of the paired-end data contained the specified adapter sequence, representing 4.0% of the total read 1 pairs.
*   **`Read 2 with adapter: 101,301 (3.2%)`**: This indicates that 101,301 reads in the second file of the paired-end data contained the specified adapter sequence, representing 3.2% of the total read 2 pairs.
*   **`Pairs written (passing filters): 3,154,678 (100.0%)`**: This shows that all 3,154,678 read pairs passed the filtering criteria (which in this case, with the provided parameters, primarily means the adapter trimming was performed, and no reads were discarded based on length or other quality metrics not specified in the command).
*   **`Total basepairs processed: 637,244,956 bp`**: This is the total number of bases across all read pairs before trimming.
*   **`Total written (filtered): 631,397,995 bp (99.1%)`**: This is the total number of bases remaining after trimming, which is 99.1% of the original base pairs. This indicates that a small percentage of bases were removed due to adapter trimming.
*   **`Read 1: 314,515,227 bp`**: The number of base pairs remaining in Read 1 after trimming.
*   **`Read 2: 316,882,768 bp`**: The number of base pairs remaining in Read 2 after trimming.
*   The subsequent sections ("=== First read: Adapter 1 ===" and "=== Second read: Adapter 2 ===") provide a detailed breakdown of the adapter trimming for each read in the pair, including the sequence of the adapter, the number of times it was trimmed, the minimum overlap required, the number of allowed errors during matching, the nucleotide composition preceding the removed adapters, and a detailed table showing the length distribution of the removed adapter sequences and the number of errors found for each length. This detailed breakdown helps in understanding the efficiency and accuracy of the adapter trimming process.

In [8]:
# 8. Running FastQC on Trimmomatic trimmed reads
!mkdir -p /content/fastqc_trimmed_trimmomatic
!fastqc /content/trimmed/SRR531365_1_paired.fq.gz \
        /content/trimmed/SRR531365_1_unpaired.fq.gz \
        /content/trimmed/SRR531365_2_paired.fq.gz \
        /content/trimmed/SRR531365_2_unpaired.fq.gz \
        -o /content/fastqc_trimmed_trimmomatic
!ls -lh /content/fastqc_trimmed_trimmomatic

from google.colab import files
files.download("/content/fastqc_trimmed_trimmomatic/SRR531365_1_paired_fastqc.html")
files.download("/content/fastqc_trimmed_trimmomatic/SRR531365_1_unpaired_fastqc.html")
files.download("/content/fastqc_trimmed_trimmomatic/SRR531365_2_paired_fastqc.html")
files.download("/content/fastqc_trimmed_trimmomatic/SRR531365_2_unpaired_fastqc.html")

Started analysis of SRR531365_1_paired.fq.gz
Approx 5% complete for SRR531365_1_paired.fq.gz
Approx 10% complete for SRR531365_1_paired.fq.gz
Approx 15% complete for SRR531365_1_paired.fq.gz
Approx 20% complete for SRR531365_1_paired.fq.gz
Approx 25% complete for SRR531365_1_paired.fq.gz
Approx 30% complete for SRR531365_1_paired.fq.gz
Approx 35% complete for SRR531365_1_paired.fq.gz
Approx 40% complete for SRR531365_1_paired.fq.gz
Approx 45% complete for SRR531365_1_paired.fq.gz
Approx 50% complete for SRR531365_1_paired.fq.gz
Approx 55% complete for SRR531365_1_paired.fq.gz
Approx 60% complete for SRR531365_1_paired.fq.gz
Approx 65% complete for SRR531365_1_paired.fq.gz
Approx 70% complete for SRR531365_1_paired.fq.gz
Approx 75% complete for SRR531365_1_paired.fq.gz
Approx 80% complete for SRR531365_1_paired.fq.gz
Approx 85% complete for SRR531365_1_paired.fq.gz
Approx 90% complete for SRR531365_1_paired.fq.gz
Approx 95% complete for SRR531365_1_paired.fq.gz
Analysis complete for SRR

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [9]:
# 9. Running FastQC on Cutadapt trimmed reads
!mkdir -p /content/fastqc_trimmed_cutadapt
!fastqc /content/cutadapt_out/SRR531365_1_trimmed.fq.gz \
        /content/cutadapt_out/SRR531365_2_trimmed.fq.gz \
        -o /content/fastqc_trimmed_cutadapt
!ls -lh /content/fastqc_trimmed_cutadapt

from google.colab import files
files.download("/content/fastqc_trimmed_cutadapt/SRR531365_1_trimmed_fastqc.html")
files.download("/content/fastqc_trimmed_cutadapt/SRR531365_2_trimmed_fastqc.html")

Started analysis of SRR531365_1_trimmed.fq.gz
Approx 5% complete for SRR531365_1_trimmed.fq.gz
Approx 10% complete for SRR531365_1_trimmed.fq.gz
Approx 15% complete for SRR531365_1_trimmed.fq.gz
Approx 20% complete for SRR531365_1_trimmed.fq.gz
Approx 25% complete for SRR531365_1_trimmed.fq.gz
Approx 30% complete for SRR531365_1_trimmed.fq.gz
Approx 35% complete for SRR531365_1_trimmed.fq.gz
Approx 40% complete for SRR531365_1_trimmed.fq.gz
Approx 45% complete for SRR531365_1_trimmed.fq.gz
Approx 50% complete for SRR531365_1_trimmed.fq.gz
Approx 55% complete for SRR531365_1_trimmed.fq.gz
Approx 60% complete for SRR531365_1_trimmed.fq.gz
Approx 65% complete for SRR531365_1_trimmed.fq.gz
Approx 70% complete for SRR531365_1_trimmed.fq.gz
Approx 75% complete for SRR531365_1_trimmed.fq.gz
Approx 80% complete for SRR531365_1_trimmed.fq.gz
Approx 85% complete for SRR531365_1_trimmed.fq.gz
Approx 90% complete for SRR531365_1_trimmed.fq.gz
Approx 95% complete for SRR531365_1_trimmed.fq.gz
Analy

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Here is a comparison of the trimming results from Trimmomatic and Cutadapt based on the provided outputs:

**Trimmomatic Output Summary:**

*   **Input Read Pairs:** 3,154,678
*   **Both Surviving:** 482,081 (15.28%) - These are the read pairs where both reads survived the trimming process and were kept as a pair.
*   **Forward Only Surviving:** 137,077 (4.35%) - These are the read pairs where only the forward read survived trimming.
*   **Reverse Only Surviving:** 15,159 (0.48%) - These are the read pairs where only the reverse read survived trimming.
*   **Dropped:** 2,520,361 (79.89%) - These are the read pairs that were completely discarded by Trimmomatic.

**Cutadapt Output Summary:**

*   **Total read pairs processed:** 3,154,678
*   **Read 1 with adapter:** 125,936 (4.0%) - Number of Read 1 reads where an adapter was found and trimmed.
*   **Read 2 with adapter:** 101,301 (3.2%) - Number of Read 2 reads where an adapter was found and trimmed.
*   **Pairs written (passing filters):** 3,154,678 (100.0%) - All read pairs were written to output files after trimming.
*   **Total basepairs processed:** 637,244,956 bp
*   **Total written (filtered):** 631,397,995 bp (99.1%) - Total base pairs remaining after trimming.

**Comparison:**

*   **Read Retention:** The most significant difference is in the number of read pairs retained. Trimmomatic discarded a large percentage of read pairs (79.89%), while Cutadapt retained all read pairs (100.0%). This suggests that Trimmomatic's filtering criteria (LEADING:3, TRAILING:3, SLIDINGWINDOW:4:15, MINLEN:36) were more stringent, leading to more reads being dropped. Cutadapt, with the parameters used, primarily performed adapter trimming and did not discard reads based on quality or length as Trimmomatic did.
*   **Adapter Trimming:** Both tools identified and trimmed adapter sequences. Cutadapt reported that adapters were found in 4.0% of Read 1 and 3.2% of Read 2. Trimmomatic's output doesn't explicitly state the percentage of reads with adapters, but the "Both Surviving", "Forward Only Surviving", and "Reverse Only Surviving" counts, along with the "Dropped" count, are a result of adapter trimming *and* other quality filtering steps.
*   **Output Files:** Trimmomatic produced four output files for paired-end data (paired forward, unpaired forward, paired reverse, unpaired reverse), reflecting its ability to handle reads where only one of the pair survives trimming. Cutadapt, with the parameters used, produced two output files for paired-end data (trimmed forward and trimmed reverse), indicating that it kept all pairs together after trimming.
*   **Focus:** Trimmomatic is a more comprehensive trimmer that can perform various quality trimming steps in addition to adapter removal. Cutadapt is primarily focused on adapter trimming and filtering based on length and quality, but the command used here only specified adapter trimming.

In summary, Trimmomatic applied more aggressive quality and length filtering, resulting in a higher number of dropped reads and the separation of surviving paired and unpaired reads. Cutadapt, with the parameters used, focused solely on adapter trimming and retained all read pairs, only reducing the total number of base pairs due to the removal of adapter sequences. The choice of which tool and parameters to use depends on the specific goals of the sequencing analysis and the desired level of quality control.